tests/test_pickling.py

```diff
@@ -0,0 +1,172 @@
+import pickle
+from copy import (
+    copy,
+    deepcopy,
+)
+from dataclasses import dataclass
+from typing import (
+    List,
+    Dict,
+)
+from unittest.mock import ANY
+
+import betterproto
+from betterproto.lib.google import protobuf as google
+
+
+def unpickled(message):
+    return pickle.loads(pickle.dumps(message))
+
+
+@dataclass(eq=False, repr=False)
+class Fe(betterproto.Message):
+    abc: str = betterproto.string_field(1)
+
+
+@dataclass(eq=False, repr=False)
+class Fi(betterproto.Message):
+    abc: str = betterproto.string_field(1)
+
+
+@dataclass(eq=False, repr=False)
+class Fo(betterproto.Message):
+    abc: str = betterproto.string_field(1)
+
+
+@dataclass(eq=False, repr=False)
+class NestedData(betterproto.Message):
+    struct_foo: Dict[str, "google.Struct"] = betterproto.map_field(
+        1, betterproto.TYPE_STRING, betterproto.TYPE_MESSAGE
+    )
+    map_str_any_bar: Dict[str, "google.Any"] = betterproto.map_field(
+        2, betterproto.TYPE_STRING, betterproto.TYPE_MESSAGE
+    )
+
+
+@dataclass(eq=False, repr=False)
+class Complex(betterproto.Message):
+    foo_str: str = betterproto.string_field(1)
+    fe: "Fe" = betterproto.message_field(3, group="grp")
+    fi: "Fi" = betterproto.message_field(4, group="grp")
+    fo: "Fo" = betterproto.message_field(5, group="grp")
+    nested_data: "NestedData" = betterproto.message_field(6)
+    mapping: Dict[str, "google.Any"] = betterproto.map_field(
+        7, betterproto.TYPE_STRING, betterproto.TYPE_MESSAGE
+    )
+
+
+def test_pickling_complex_message():
+    msg = Complex(
+        foo_str="yep",
+        fe=Fe(abc="1"),
+        nested_data=NestedData(
+            struct_foo={
+                "foo": google.Struct(
+                    fields={
+                        "hello": google.Value(
+                            list_value=google.ListValue(
+                                values=[google.Value(string_value="world")]
+                            )
+                        )
+                    }
+                ),
+            },
+            map_str_any_bar={
+                "key": google.Any(value=b"value"),
+            },
+        ),
+        mapping={
+            "message": google.Any(value=bytes(Fi(abc="hi"))),
+            "string": google.Any(value=b"howdy"),
+        },
+    )
+    deser = unpickled(msg)
+    assert msg == deser
+    assert msg.fe.abc == "1"
+    assert msg.is_set("fi") is not True
+    assert msg.mapping["message"] == google.Any(value=bytes(Fi(abc="hi")))
```

Comment:
See this assertion and the following two lines

Should fields be perfectly preserved in this manner? My gut instinct is yes... but it may be annoying for users to extract and decode these values. It is probably not the job of this library to take care of that.

Reply:
I think it's maybe something to worry about after Any support, for now this seems fine